tree_in_context/src/interpreter.rs

```diff
@@ -1,6 +1,14 @@
-use std::{collections::HashMap, fmt::Display};
+//! Interpret DAG programs. Each expression is evaluated once, so side-effects only happen
+//! once. All the dependencies of an expression are evaluted first.
+//! The interpreter relies on the invariant that common subexpressions are
```

Comment:
Is this invariant checked/enforced or is this comment the only place it's clearly documented?